tests/scenario/test_ingress_per_app.py

```diff
@@ -0,0 +1,172 @@
+import yaml
+from ops import pebble
+from scenario import Container, Context, Model, Mount, Relation, State
+
+
+@pytest.fixture
+def context(traefik_charm):
+    return Context(charm_type=traefik_charm)
+
+
+@pytest.fixture
+def model():
+    return Model(name="test-model")
+
+
+@pytest.fixture
+def temp_opt():
+    with tempfile.TemporaryDirectory() as td:
+        yield Path(td)
+
+
+@pytest.fixture
+def traefik_container(temp_opt):
+    layer = pebble.Layer(
+        {
+            "summary": "Traefik layer",
+            "description": "Pebble config layer for Traefik",
+            "services": {
+                "traefik": {
+                    "override": "replace",
+                    "summary": "Traefik",
+                    "command": '/bin/sh -c "/usr/bin/traefik | tee /var/log/traefik.log"',
+                    "startup": "enabled",
+                },
+            },
+        }
+    )
+
+    opt = Mount("/opt/", temp_opt)
+
+    return Container(
+        name="traefik",
+        can_connect=True,
+        layers={"traefik": layer},
+        service_status={"traefik": pebble.ServiceStatus.ACTIVE},
+        mounts={"opt": opt},
+    )
+
+
+@pytest.mark.parametrize("port, host", ((80, "1.1.1.1"), (81, "10.1.10.1")))
+@pytest.mark.parametrize("event_name", ("joined", "changed"))
+def test_ingress_per_app_created(
+    context, port, host, model, traefik_container, event_name, temp_opt
+):
+    """Check the config when a new ingress per leader is created or changes (single remote unit)."""
+    ipl = Relation(
+        "ingress",
+        remote_app_data={
+            "model": "test-model",
+            "name": "remote/0",
+            "mode": "http",
+        },
+        remote_units_data={0: {"port": str(port), "host": host}},
+        relation_id=1,
+    )
+    state = State(
+        model=model,
+        config={"routing_mode": "path", "external_hostname": "foo.com"},
+        containers=[traefik_container],
+        relations=[ipl],
+    )
+
+    event = getattr(ipl, f"{event_name}_event")
+    context.run(event, state)
+
+    generated_config = yaml.safe_load(
+        traefik_container.filesystem.open(
+            "/opt/traefik/juju/juju_ingress_ingress_1_remote.yaml"
+        ).read()
+    )
+
+    assert generated_config["http"]["services"]["juju-test-model-remote-0-service"] == {
+        "loadBalancer": {"servers": [{"url": f"http://{host}:{port}"}]}
+    }
+
+
+@pytest.mark.parametrize("port, host", ((80, "1.1.1.{}"), (81, "10.1.10.{}")))
+@pytest.mark.parametrize("n_units", (2, 3, 10))
+def test_ingress_per_app_scale(context, host, port, model, traefik_container, temp_opt, n_units):
+    """Check the config when a new ingress per leader unit joins."""
+    cfg_file = temp_opt.joinpath("traefik", "juju", "juju_ingress_ingress_1_remote.yaml")
+    cfg_file.parent.mkdir(parents=True)
+
+    # config that would have been generated from mock_data_0
+    # same as config output of the previous test
+    initial_cfg = {
```

Comment:
This test concerns `...["servers"]`.

Instead of hard-coding the entire relation data, would it make more sense to have a fixture that brings the charm through an equivalent of `begin_with_inital_hooks`?

Reply:
It would, but I'm trying to keep each test isolated.
I'm sure we'll refactor this once we change the schema and realize all of the tests break because we're overfitting the data in its current form, but for now I vote to keep it as-is